File: wpilibj/src/test/java/edu/wpi/first/wpilibj/XboxControllerTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.hal.HAL;
import edu.wpi.first.wpilibj.simulation.XboxControllerSim;
import org.junit.jupiter.api.Test;

class XboxControllerTest {
  @Test
  void testGetX() {
    HAL.initialize(500, 0);
    XboxController joy = new XboxController(2);
    XboxControllerSim joysim = new XboxControllerSim(joy);

    joysim.setLeftX(0.35);
    joysim.setRightX(0.45);
    joysim.notifyNewData();
    assertEquals(0.35, joy.getLeftX(), 0.001);
    assertEquals(0.45, joy.getRightX(), 0.001);
  }

  @Test
  void testGetBumper() {
    HAL.initialize(500, 0);
    XboxController joy = new XboxController(2);
    XboxControllerSim joysim = new XboxControllerSim(joy);

    joysim.setLeftBumper(false);
    joysim.setRightBumper(true);
    joysim.notifyNewData();
    assertFalse(joy.getLeftBumper());
    assertTrue(joy.getRightBumper());
    // need to call pressed and released to clear flags
    joy.getLeftBumperPressed();
    joy.getLeftBumperReleased();
    joy.getRightBumperPressed();
    joy.getRightBumperReleased();

    joysim.setLeftBumper(true);
    joysim.setRightBumper(false);
    joysim.notifyNewData();
    assertTrue(joy.getLeftBumper());
    assertTrue(joy.getLeftBumperPressed());
    assertFalse(joy.getLeftBumperReleased());
    assertFalse(joy.getRightBumper());
    assertFalse(joy.getRightBumperPressed());
    assertTrue(joy.getRightBumperReleased());
  }

  @Test
  void testGetAButton() {
    HAL.initialize(500, 0);
    XboxController joy = new XboxController(2);
    XboxControllerSim joysim = new XboxControllerSim(joy);

    joysim.setAButton(false);
    joysim.notifyNewData();
    assertFalse(joy.getAButton());
    // need to call pressed and released to clear flags
    joy.getAButtonPressed();
    joy.getAButtonReleased();

    joysim.setAButton(true);
    joysim.notifyNewData();
    assertTrue(joy.getAButton());
    assertTrue(joy.getAButtonPressed());
    assertFalse(joy.getAButtonReleased());

    joysim.setAButton(false);
    joysim.notifyNewData();
    assertFalse(joy.getAButton());
    assertFalse(joy.getAButtonPressed());
    assertTrue(joy.getAButtonReleased());
  }
}

File: wpilibc/src/test/native/cpp/XboxControllerTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/XboxController.h"  // NOLINT(build/include_order)

#include "frc/simulation/XboxControllerSim.h"
#include "gtest/gtest.h"

using namespace frc;

TEST(XboxControllerTests, GetX) {
  XboxController joy{2};
  sim::XboxControllerSim joysim{joy};

  joysim.SetLeftX(0.35);
  joysim.SetRightX(0.45);
  joysim.NotifyNewData();
  ASSERT_NEAR(joy.GetLeftX(), 0.35, 0.001);
  ASSERT_NEAR(joy.GetRightX(), 0.45, 0.001);
}

TEST(XboxControllerTests, GetBumper) {
  XboxController joy{1};
  sim::XboxControllerSim joysim{joy};

  joysim.SetLeftBumper(false);
  joysim.SetRightBumper(true);
  joysim.NotifyNewData();
  ASSERT_FALSE(joy.GetLeftBumper());
  ASSERT_TRUE(joy.GetRightBumper());
  // need to call pressed and released to clear flags
  joy.GetLeftBumperPressed();
  joy.GetLeftBumperReleased();
  joy.GetRightBumperPressed();
  joy.GetRightBumperReleased();

  joysim.SetLeftBumper(true);
  joysim.SetRightBumper(false);
  joysim.NotifyNewData();
  ASSERT_TRUE(joy.GetLeftBumper());
  ASSERT_TRUE(joy.GetLeftBumperPressed());
  ASSERT_FALSE(joy.GetLeftBumperReleased());
  ASSERT_FALSE(joy.GetRightBumper());
  ASSERT_FALSE(joy.GetRightBumperPressed());
  ASSERT_TRUE(joy.GetRightBumperReleased());
}

TEST(XboxControllerTests, GetAButton) {
  XboxController joy{1};
  sim::XboxControllerSim joysim{joy};

  joysim.SetAButton(false);
  joysim.NotifyNewData();
  ASSERT_FALSE(joy.GetAButton());
  // need to call pressed and released to clear flags
  joy.GetAButtonPressed();
  joy.GetAButtonReleased();

  joysim.SetAButton(true);
  joysim.NotifyNewData();
  ASSERT_TRUE(joy.GetAButton());
  ASSERT_TRUE(joy.GetAButtonPressed());
  ASSERT_FALSE(joy.GetAButtonReleased());

  joysim.SetAButton(false);
  joysim.NotifyNewData();
  ASSERT_FALSE(joy.GetAButton());
  ASSERT_FALSE(joy.GetAButtonPressed());
  ASSERT_TRUE(joy.GetAButtonReleased());
}
